### PR TITLE
add factory for ASF service providers

### PR DESCRIPTION
**Pull request by thrau**
This PR adds a factory for creating `Service` instances for ASF providers.

Through the `MotoFallbackDispatcher` we lost the reference the original provider instance. Basically there was no way to use the `ServicePluginManager` to iterate over the provider instances. This is however necessary for the state visitor pattern to work correctly. Providers need to be able to implement the `accept_state_visitor` method, and some higher-order mechanism needs to collect all provider instances and call the method on them. Now this can be the `Service` instance, since the `Service` can delegate the call to the `_provider` attribute it now holds.
So basically what this will enable down the road is

```python
for service in SERVICE_PLUGINS:
     service.accept_state_visitor(visitor)
```

It also creates a nice way of phasing out the `AwsApiListener` which will go away soon :partying_face:  it isolates the use into the factory method.

Plus it makes the `providers.py` a bit cleaner :-)

### CODE REVIEW

File: localstack/services/plugins.py [hunk 1]
```diff
@@ -220,6 +234,44 @@ def name(self):
...
+    ) -> "Service":
+        """
+        Factory method for creating services for providers. This method hides a bunch of legacy code and
+        band-aids/adapters to make persistence visitors work, while providing compatibility with the legacy edge proxy.
+
+        :param provider: the service provider, i.e., the implementation of the generated ASF service API.
+        :param dispatch_table_factory: a `MotoFallbackDispatcher` or something similar that uses the provider to
+            create a dispatch table. this one's a bit clumsy.
+        :param service_lifecycle_hook: if left empty, then it checks whether the provider is a ServiceLifecycleHook.
+            if set to None, it will disable the lifecyle hook alltogether
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/plugins.py:251]: no not used anywhere yet, just added it for flexibility. can remove it

**Comment by thrau** [comment]
thanks @alexrashed , i think those are all great points, but out of scope of this PR. we are definitely going to iterate over the Service concept in the near future, so everything will look different anyway :-)

File: localstack/services/plugins.py [hunk 2]
```diff
@@ -220,6 +234,44 @@ def name(self):
     def is_enabled(self):
         return True
 
+    @staticmethod
+    def for_provider(
+        provider: ServiceProvider,
+        dispatch_table_factory: Callable[[ServiceProvider], DispatchTable] = None,
```
[the remark this answers is not in the document]
Comment (thrau) [reply localstack/services/plugins.py:240]: i agree with the coupling (as we discussed previously)

while working on this PR i was playing around with something like a `MotoFallbackDispatcherMixin` that the provider could inherit from, and that would perform the class instrumentation. but some new issues surfaced with that solution so i put it on hold for now.

> having the service provider itself create the dispatcher

i'm not sure how that would look like exactly without introducing the "dispatcher" concept on provider-level (which i don't think we should do - providers don't and shouldn't know anything about dispatching). the mixing approach i described above (which i think is the way to go), needs a bit of tinkering, but can be a good compromise.

anway, TBD

**Review by giograno — APPROVED** [review]
As for the scope of the PR, it all looks good! Like the factory and the fact that now we have a `ServiceProvider` protocol.

Outside the scope of the PR, just for myself to start ironing this thing out. 
I am convinced that providers should hold all the logic for before/after `<put_your_event_here>` and that we don't need visitors for it. Furthermore, If I understood correctly: i) the providers will be the new `StateVisitable`; ii) they hold a state locator (that is now the `StateVisitable`); and iii) dispatch the `visit` via such a state locator.
